growcrops.lua

```diff
@@ -0,0 +1,138 @@
+
+if not material or material == "help" or material == "list" then
+    listPlantedCrops(crop_raws, inventory)
+else
+    if material == 'all' then
+        for mat_idx, _ in pairs(inventory) do
+            growCrop(crop_raws, crop_raws[mat_idx].name)
+        end
+    else
+        growCrop(crop_raws, material)
+    end
+end
+
+-- Collects crop name and growdur from the raws.
+-- The returned table can be indexed by mat_index
+function cacheCropRaws()
+    local crops = {}
+    -- store ID and grow duration for each crop
+    for idx, p in ipairs(df.global.world.raws.plants.all) do
+        crops[idx] = { name = p.id, growdur = p.growdur }
+    end
+
+    return crops
+end
+
+-- determine if the given refs collection contains the type
+-- e.g. containsRefType(item.general_refs, df.general_ref_building_holderst)
+function containsRefType(refs, type)
+    for i, ref in ipairs(refs) do
+        if ref._type == type then
+            return true
+        end
+    end
+
+    return false
+end
+
+-- create a table of planted seed counts indexed by mat_index
+function collectInventory(crop_raws)
+    local inventory = {}
+    for _, seed in ipairs(df.global.world.items.other.SEEDS) do
+        -- skip if seed is not in a specific building (plot)
+        if not containsRefType(seed.general_refs, df.general_ref_building_holderst) then
+            goto skipseed
+        end
+
+        if seed.grow_counter < crop_raws[seed.mat_index].growdur then
+            inventory[seed.mat_index] = (inventory[seed.mat_index] or 0) + 1
+        end
+
+        :: skipseed ::
+    end
+
+    return inventory
+end
+
+-- Display a list of planted crops
+function listPlantedCrops(crop_raws, inventory)
+    local crops = {}
+    -- create a copy of crops from inventory to use for sorting
+    for mat, invCount in pairs(inventory) do
+        table.insert(crops, { name = crop_raws[mat].name, count = invCount })
+    end
+
+    -- sort the table in descending order by seed count
+    table.sort(crops, function(a, b) return a.count > b.count end)
+
+    for _, crop in ipairs(crops) do
+        print(string.format("%s %d", crop.name, crop.count))
+    end
+end
+
+-- Tries to find a crop raw in crop_raws that matches the given name
+-- The returned value contains the mat_index and name of the crop
+function findRawsCrop(crop_raws, material)
+    if not material then
+        return nil
+    end
+
+    local possible_matches = {}
+    local lower_mat = string.lower(material)
+    for mat_index, crop in pairs(crop_raws) do
+        local lower_crop_name = string.lower(crop.name)
+        if lower_mat == lower_crop_name then
+            return {mat_index = mat_index, name = crop.name}
+        end
+
+        if string.find(lower_crop_name, lower_mat) then
+            table.insert(possible_matches, {mat_index = mat_index, name = crop.name})
+        end
+    end
+
+    if #possible_matches == 1 then
+        return possible_matches[1]
+    end
```

Comment:
Is there a better way to do this with the lua api?

The ruby script was using `df.match_rawname` but that doesn't appear to exist for the lua api. So I basically inline implemented that function here

Reply:
I'm not aware of any function that does this in lua, though if this comes up elsewhere I'll recommend that this be made into a library function.

Reply:
I will at least pull this implementation out into a function (not really sure why I didn't do that to begin with....) so that it can easily be used by others or easily pulled to somewhere more common in the future